### PR TITLE
[Core] Add object-factory option to CLI and CucumberOptions

### PR DESCRIPTION
**Pull request by ralphkar**
## Summary

In order to provide different test configurations it might be useful to use different object factories when working with a DI framework. This is described in issue #1690.

## Details

- The annotations for JUnit and TestNG now allow to specify the class to instantiate as object factory.
- A CLI flag has been added to allow the specification of an object factory on the command line (--object-factory).
- Some minor refactorings along the way.

## Motivation and Context

- It is now possible to use different object factories in a Cucumber test suite. Features need to be separated (using tags) and can be grouped as separate tests that are annotated with different CucumberOptions.
- For each specification of CucumberOptions a different objectFactory can be specified. That way different feature tests can e.g. use differently configured injectors.
- It is no longer necessary to specify an object factory in `cucumber.properties`.
- Object factories are loaded via the ServiceLoader and need to be specified in the file `META-INF/services/io.cucumber.core.backend.ObjectFactory`. Multiple implementations can be given here.

## How Has This Been Tested?

Unit tests have been added and a production test has been executed with our product.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue).
- [x] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).

## Checklist:

- [x] I've added tests for my code.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: core/src/main/java/io/cucumber/core/options/RuntimeOptionsParser.java [hunk 1]
```diff
@@ -105,6 +105,9 @@ RuntimeOptionsBuilder parse(List<String> args) {
                     throw new CucumberException("--count must be > 0");
                 }
                 parsedOptions.setCount(count);
+            } else if (arg.equals("--object-factory")) {
+                String objectFactoryClassName = args.remove(0).trim();
```
Comment (mpkorstanje) [inline core/src/main/java/io/cucumber/core/options/RuntimeOptionsParser.java]: Is the trim needed?

File: junit/src/main/java/io/cucumber/junit/JUnitCucumberOptionsProvider.java [hunk 2]
```diff
@@ -67,14 +68,12 @@ public boolean monochrome() {
...
         public SnippetType snippets() {
-            switch (annotation.snippets()) {
-                case UNDERSCORE:
-                    return SnippetType.UNDERSCORE;
-                case CAMELCASE:
-                    return SnippetType.CAMELCASE;
-                default:
-                    throw new IllegalArgumentException("" + annotation.snippets());
-            }
+            return annotation.snippets();
```
Comment (mpkorstanje) [inline junit/src/main/java/io/cucumber/junit/JUnitCucumberOptionsProvider.java]: Why did you also change this?

File: core/src/test/java/io/cucumber/core/options/CucumberPropertiesParserTest.java [hunk 3]
```diff
@@ -44,14 +43,9 @@ public boolean addClass(Class<?> glueClass) {
         }
 
         @Override
-        public void start() {
-
-        }
+        public void start() {}
```
Comment (mpkorstanje) [inline core/src/test/java/io/cucumber/core/options/CucumberPropertiesParserTest.java]: Please keep these separated. The next auto format with IDEA will just revert it.

File: junit/src/main/java/io/cucumber/junit/JUnitCucumberOptionsProvider.java (hunk 2 above)
Comment (mpkorstanje) [reply junit/src/main/java/io/cucumber/junit/JUnitCucumberOptionsProvider.java]: Ah. I see. `io.cucumber.core.snippets.SnippetType` is still part of the public API. It shouldn't be. The local duplicates keep these sort of implementation details away from the user.

Comment (mpkorstanje) [reply junit/src/main/java/io/cucumber/junit/JUnitCucumberOptionsProvider.java]: Can you instead revert this change and remove the `@API` annotation from `io.cucumber.core.snippets.SnippetType`?

**Comment by mpkorstanje** [comment]
Cheers looks good. Just a few small things.

File: core/src/test/java/io/cucumber/core/options/CucumberPropertiesParserTest.java (hunk 3 above)
Comment (ralphkar) [reply core/src/test/java/io/cucumber/core/options/CucumberPropertiesParserTest.java]: Done

File: core/src/main/java/io/cucumber/core/options/RuntimeOptionsParser.java (hunk 1 above)
Comment (ralphkar) [reply core/src/main/java/io/cucumber/core/options/RuntimeOptionsParser.java]: Probably not ... I removed it. Just trying to be overly safe ;-)

File: junit/src/main/java/io/cucumber/junit/JUnitCucumberOptionsProvider.java (hunk 2 above)
Comment (ralphkar) [reply junit/src/main/java/io/cucumber/junit/JUnitCucumberOptionsProvider.java]: Not sure why this is better. Anyway, I reverted the changes for the JUnit and TestNG CucumberOptions annotations (the part about snippets) and removed the `@API` annotation from `io.cucumber.core.snippets.SnippetType`.

**Comment by ralphkar** [comment]
I have made the requested changes. I think the PR is ready. Maybe at some point in the future someone should write a short abstract on how to use object factories in v5.

File: junit/src/main/java/io/cucumber/junit/JUnitCucumberOptionsProvider.java (hunk 2 above)
Comment (mpkorstanje) [reply junit/src/main/java/io/cucumber/junit/JUnitCucumberOptionsProvider.java]: Cheers.

Currently a user needs to declare two dependencies to user cucumber. For example `cucumber-java` and `cucumber-junit`. The transitive dependencies are automatically included by dependency management tools. 

When designing a library it is a good practice not to expose classes from transitive dependencies to the consumer.  This makes versioning easier. `SnippetType` is defined in `core`. By having an alias in `cucumber-junit` we avoid exposing core.

**Comment by mpkorstanje** [comment]
Cheers!

> Maybe at some point in the future someone should write a short abstract on how to use object factories in v5.

You wouldn't be happen to volunteer? It doesn't need to be a full blown doc. A short brief in a readme of the core module will do.

**Comment by ralphkar** [comment]
> > Maybe at some point in the future someone should write a short abstract on how to use object factories in v5.
> 
> You wouldn't be happen to volunteer? It doesn't need to be a full blown doc. A short brief in a readme of the core module will do.

As requested -> #1715

**Pull request merged**